x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
         }
     }
 
-    private void readInternalBytes(final long part, final long pos, byte[] b, int offset, int length) throws IOException {
+    private void readInternalBytes(final int part, long pos, final byte[] b, int offset, int length) throws IOException {
+        final long currentSequentialReadSize = sequentialReadSize;
+        if (currentSequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            final StreamForSequentialReads streamForSequentialReads = streamForSequentialReadsRef.get();
+            if (streamForSequentialReads == null) {
+                // start a new sequential read
+                if (tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                    return;
+                }
+            } else if (streamForSequentialReads.part == part && streamForSequentialReads.pos == pos) {
+                // continuing a sequential read that we started previously
+                assert streamForSequentialReads.isFullyRead() == false;
+                int read = streamForSequentialReads.inputStream.read(b, offset, length);
+                assert read <= length : read + " vs " + length;
+                streamForSequentialReads.pos += read;
+                position += read;
+                pos += read;
+                offset += read;
+                length -= read;
+
+                if (streamForSequentialReads.isFullyRead()) {
+                    if (streamForSequentialReadsRef.compareAndSet(streamForSequentialReads, null) != false) {
+                        streamForSequentialReads.close();
+                    } else {
+                        // something happened concurrently, defensively stop optimizing
+                        sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
```

Comment:
Do we ever expect concurrent activity here? If that was the case, our bookkeeping logic (position, pos, offset, length) would not work at all?

Reply:
We don't expect concurrent reading, but we can be concurrently closed. In an earlier iteration we also dropped the optimisation if cloned or sliced, which happens on a separate thread, but in fact that shouldn't be happening concurrently with reading either, so concurrent closing is the only thing to handle. I'll adjust the comments.

Reply:
Adjusted in 84d12b606f06c3d37e2dd9bb24549356079d913a.